doc/zk-integration.md

```diff
@@ -0,0 +1,214 @@
+ which corresponds to a mode where the rollup falls back to using its own (centralized) sequencer as shown in Figure 1.
+
+![image](zk-rollup-default-sequencer.svg)
+
+**Figure 1:** Circuit used when the rollup falls back to using its default (centralized) sequencer. The public inputs
+are the current state of the rollup _COMM_STATE_VM i_, the new rollup state after update _COMM_STATE_VM i+1_ and a
+commitment to the transactions applied to the state, _COMM_TXS_ROLLUP_. The private input (in bold) corresponds the list
+of transactions.
+
+## Integration 1: Rollup contract fetches Espresso block commitment from the Espresso light client contract
+
+For this integration, Espresso consensus verification is delegated to the Espresso light client contract. In practice
+the rollup contract will fetch the last Espresso block commitment and feed it to the circuit. Still additional gadgets
+need to be introduced in order to implement the derivation pipeline logic consisting at a high level in:
+
+- Collecting all the Espresso commitments since the last update.
+- For each of these commitments, filter the corresponding Espresso blocks in order to obtain the transactions belonging
+  to the rollups.
+- Establishing some equivalence between the commitment to the rollup transactions used in the zkVM and the commitment to
+  those same transactions obtained after namespace filtering.
+
+![image](zk-rollup-circuit-no-espresso-consensus.svg) **Figure 2:** Rollup circuit with additional gadgets _Collect &
+Filter_, and _COMMs Equivalence_. Private inputs of the circuit are written in bold font.
+
+The circuit depicted in Figure 2 operates as follows:
+
+- The _Collect & Filter_ gadget receives as input _BLOCK_COMM_ESP_NEW_ which is the commitment to the latest Espresso
+  block available and _BLOCK_COMM_ESP_OLD_. Both of these commitments are public inputs. The first witness of this
+  circuit is _COMM_TXS_HISTORY_ESP_ which is a commitment to all the rollup transactions that have been sequenced since
+  the last Espresso state update _BLOCK_COMM_ESP_OLD_. The relationship between _BLOCK_COMM_ESP_, _BLOCK_COMM_ESP_OLD_,
+  and _COMM_TXS_HISTORY_ESP_ can be checked using a second witness _PROOF_TXS_HISTORY_ESP_.
+- The _COMMs Equivalence_ gadget checks that using the same rollup inputs _ROLLUP_TXS_, we obtain _COMM_TXS_HISTORY_ESP_
+  using the Espresso commitment scheme for representing a set of transactions and the commitment _COMM_TXS_ROLLUPS_ that
+  is used the by _zkVM_ gadget.
+- The _zkVM_ gadget is the original gadget of the rollup circuit that proves a correct transition from state
+  _COMM_STATE_VM i_ to the next state _COMM_STATE_VM i+1_ when applying the transactions represented by the commitment
+  value _COMM_TXS_ROLLUP_.
+- These three gadgets above return a boolean: true if the verification succeeds and false otherwise.
+- For the circuit to accept, all these gadget outputs must be true, and thus we add an _AND_ gate.
+
+```solidity
```

Comment:
Please add a caption explaining why this solidity code is here, similar to the captions for previous diagrams.

Reply:
Improved the description of the rollup contracts in e246dc999fc8c17fa98f25db7026e1945f1a3aa2.